### PR TITLE
use absolute urls throughout reports for better emails

### PR DESCRIPTION
**Pull request by dannyroberts**
http://manage.dimagi.com/default.asp?154859

merge also https://github.com/dimagi/couchlog/pull/16 and https://github.com/dimagi/dimagi-utils/pull/246

### CODE REVIEW

File: corehq/apps/reports/display.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 import json
 from couchdbkit.exceptions import MultipleResultsFound, NoResultFound
+from corehq.util.view_utils import absolute_reverse
```
Comment (dimagimon) [inline corehq/apps/reports/display.py]: F401 'absolute_reverse' imported but unused

File: corehq/apps/reports/generic.py [hunk 2]
```diff
@@ -645,7 +646,7 @@ def get_url(cls, domain=None, render_as=None, **kwargs):
         url_args = [domain] if domain is not None else []
         if render_as is not None:
             url_args.append(render_as+'/')
-        return reverse(cls.dispatcher.name(), args=url_args+[cls.slug])
+        return absolute_reverse(cls.dispatcher.name(), args=url_args+[cls.slug])
```
Comment (dimagimon) [inline corehq/apps/reports/generic.py]: E226 missing whitespace around arithmetic operator

**Comment by dannyroberts** [comment]
@esoergel or others, let me know if you see any general problems with this approach

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by czue** [comment]
looks good to me.

**Comment by dannyroberts** [comment]
I did some basic local testing and deployed to staging last night. The change touches a lot of places, so other than being confident in my own diligence and logic when refactoring, I'm not sure it's worth trying to test every single usage. I tested the functional change I was going for, buried in here, which is that email reports should use absolute urls.

**Comment by dannyroberts** [comment]
@czue good to go from my end

File: corehq/util/view_utils.py [hunk 3]
```diff
@@ -114,3 +117,7 @@ def get_request():
                 return frame.f_locals[code.co_varnames[0]]
     finally:
         del frame
+
+
+def absolute_reverse(*args, **kwargs):
+    return "{}{}".format(get_url_base(), reverse(*args, **kwargs))
```
Comment (esoergel) [inline corehq/util/view_utils.py]: :shipit:

**Comment by esoergel** [comment]
Only edge case I can think of offhand is commtrack.org, which it looks like you covered.  I assume you took that into account?

**Comment by dannyroberts** [comment]
We talked through that, and I think it just means when you click through a link in a report, it'll take you to www.commcarehq.org regardless of your original DNS name

**Pull request merged**